File: WorBreak.py
"""
Basic DSA - Word Break
Return True if s can be segmented into a space-separated sequence of one or more dictionary words.
"""


def word_break(s, word_list):
    """
    DP: dp[i] means s[:i] can be segmented.
    Time Complexity: O(n^3) naive substring checks; word set lookup is O(1)
    Space Complexity: O(n)
    """
    words = set(word_list)
    n = len(s)
    dp = [False] * (n + 1)
    dp[0] = True

    for i in range(1, n + 1):
        for j in range(i):
            if dp[j] and s[j:i] in words:
                dp[i] = True
                break

    return dp[n]


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Word Break")
    print("=" * 60)

    tests = [
        ("leetcode", ["leet", "code"], True),
        ("applepenapple", ["apple", "pen"], True),
        ("catsandog", ["cats", "dog", "sand", "and", "cat"], False),
        ("", ["a"], True),
    ]

    for s, words, expected in tests:
        got = word_break(s, words)
        print(f"s={s!r} -> {got} (expected {expected})")

File: WorBreak.java
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Basic DSA Problems - One Fundamental Dynamic Programming Question
 * Simple and essential DSA problem for beginners
 */
public class WorBreak {

    // Problem 1: Word Break
    // Can s be split into spaced words, each appearing in wordDict (reuse allowed)?

    /**
     * dp[i] = true iff prefix s[0..i) can be segmented
     * Time Complexity: O(n^2 * n) substring = O(n^3) naive; acceptable for small n in demos
     * Space Complexity: O(n)
     */
    public static Boolean wordBreak(String s, List<String> wordDict) {
        if (s == null || wordDict == null) {
            return null;
        }

        Set<String> words = new HashSet<>(wordDict);
        int n = s.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true;

        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && words.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[n];
    }

    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Word Break");
        System.out.println("=".repeat(60));

        System.out.println(
                "leetcode -> "
                        + wordBreak("leetcode", List.of("leet", "code")));
        System.out.println(
                "applepenapple -> "
                        + wordBreak("applepenapple", List.of("apple", "pen")));
        System.out.println(
                "catsandog -> "
                        + wordBreak("catsandog", List.of("cats", "dog", "sand", "and", "cat")));
        System.out.println("null s -> " + wordBreak(null, List.of("a")));
    }
}